src/index.d.ts

```diff
@@ -0,0 +1,1107 @@
+    prop1: P1,
+    prop2: P2,
+    prop3: P3,
+    prop4: P4
+  ): void;
+  each(
+    callback: EachCallback<any[]>,
+    prop1: PropertyKey,
+    prop2: PropertyKey,
+    prop3: PropertyKey,
+    prop4: PropertyKey,
+    ...props: PropertyKey[]
+  ): void;
+
+  map<U>(callback: MapCallback<T, U>): U[];
+  map<U, P1 extends IterProperties<T>>(
+    callback: MapCallback<IndexValue<T, P1>, U>,
+    prop1: P1
+  ): U[];
+  map<U, P1 extends keyof T, P2 extends IterProperties<T[P1]>>(
+    callback: MapCallback<IndexValue<IndexValue<T, P1>, P2>, U>,
+    prop1: P1,
+    prop2: P2
+  ): U[];
+  map<
+    U,
+    P1 extends keyof T,
+    P2 extends IterProperties<T[P1]>,
+    P3 extends IterProperties<IndexValue<T[P1], P2>>
+  >(
+    callback: MapCallback<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>, U>,
+    prop1: P1,
+    prop2: P2,
+    prop3: P3
+  ): U[];
+  map<
+    U,
+    P1 extends keyof T,
+    P2 extends IterProperties<T[P1]>,
+    P3 extends IterProperties<IndexValue<T[P1], P2>>,
+    P4 extends IterProperties<IndexValue<IndexValue<T[P1], P2>, P3>>
+  >(
+    callback: MapCallback<
+      IndexValue<IndexValue<IndexValue<IndexValue<T, P1>, P2>, P3>, P4>,
+      U
+    >,
+    prop1: P1,
+    prop2: P2,
+    prop3: P3,
+    prop4: P4
+  ): U[];
+  map<U>(
+    callback: MapCallback<any[], U>,
+    prop1: PropertyKey,
+    prop2: PropertyKey,
+    prop3: PropertyKey,
+    prop4: PropertyKey,
+    ...props: PropertyKey[]
+  ): U[];
+}
+
+/** @deprecated `FastPath` was renamed to `AstPath` */
+export type FastPath<T = any> = AstPath<T>;
+
+export type BuiltInParser = (text: string, options?: any) => AST;
+export type BuiltInParserName =
+  | "acorn"
+  | "angular"
+  | "babel-flow"
+  | "babel-ts"
+  | "babel"
+  | "css"
+  | "espree"
+  | "flow"
+  | "glimmer"
+  | "graphql"
+  | "html"
+  | "json-stringify"
+  | "json"
+  | "json5"
+  | "less"
+  | "lwc"
+  | "markdown"
+  | "mdx"
+  | "meriyah"
+  | "scss"
+  | "typescript"
+  | "vue"
+  | "yaml";
+export type BuiltInParsers = Record<BuiltInParserName, BuiltInParser>;
+
+export type CustomParser = (
+  text: string,
+  options: Options
+) => AST | Promise<AST>;
+
+/**
+ * For use in `.prettierrc.js`, `.prettierrc.cjs`, `prettierrc.mjs`, `prettier.config.js`, `prettier.config.cjs`, `prettier.config.mjs`
+ */
+export interface Config extends Options {
+  overrides?: Array<{
+    files: string | string[];
+    excludeFiles?: string | string[];
+    options?: Options;
+  }>;
+}
+
+export interface Options extends Partial<RequiredOptions> {}
+
+export interface RequiredOptions extends doc.printer.Options {
+  /**
+   * Print semicolons at the ends of statements.
+   * @default true
+   */
+  semi: boolean;
+  /**
+   * Use single quotes instead of double quotes.
+   * @default false
+   */
+  singleQuote: boolean;
+  /**
+   * Use single quotes in JSX.
+   * @default false
+   */
+  jsxSingleQuote: boolean;
+  /**
+   * Print trailing commas wherever possible.
+   * @default 'all'
```

Comment:
Let's use double quote? We default to double quote, though we don't format JSDoc.